tests/test_util.py

```diff
@@ -107,6 +123,160 @@ def test_clean_stale_shared_memory():
+        1. All urls are str (local). All urls are accessible locally -> no download
+        2. All urls are str (local). At least one url is unaccessible locally -> Error
+        3. All urls are tuple (local, remote). All urls are accessible locally -> no download
+        4. All urls are tuple (local, remote). At least one url is not accessible locally -> download all
+        5. All urls are str (remote) -> download all
+    """
+    from decimal import Decimal
+
+    from pyspark.sql import SparkSession
+    from pyspark.sql.types import DecimalType, IntegerType, StringType, StructField, StructType
+
+    from streaming.base.converters import dataframeToMDS
+
+    def not_merged_index(index_file_path: str, out: str):
+        """Check if index_file_path is the merged index at folder out."""
+        prefix = str(urllib.parse.urlparse(out).path)
+        return os.path.dirname(index_file_path).strip('/') != prefix.strip('/')
+
+    local, _ = local_remote_dir
+
+    mds_out = out = local
+    scheme = 's3://'
+
+    spark = SparkSession.builder.getOrCreate()  # pyright: ignore
+    schema = StructType([
+        StructField('id', IntegerType(), nullable=False),
+        StructField('name', StringType(), nullable=False),
+        StructField('amount', DecimalType(10, 2), nullable=False)
+    ])
+    data = [(1, 'Alice', Decimal('123.45')), (2, 'Bob', Decimal('67.89')),
+            (3, 'Charlie', Decimal('987.65'))]
+    df = spark.createDataFrame(data=data, schema=schema).repartition(3)
+    mds_kwargs = {'out': mds_out, 'columns': {'id': 'int', 'name': 'str'}, 'keep_local': True}
+    dataframeToMDS(df, merge_index=False, mds_kwargs=mds_kwargs)
+
+    local_cu = CloudUploader.get(local, exist_ok=True, keep_local=True)
+    local_index_files = [
+        o for o in local_cu.list_objects() if o.endswith('.json') and not_merged_index(o, local)
+    ]
+
+    if index_file_urls_pattern == 1:
+        merge_index(local_index_files, out, keep_local=keep_local)
+
+    if index_file_urls_pattern == 2:
+        with tempfile.TemporaryDirectory() as a_temporary_folder:
+            index_file_urls = [
+                os.path.join(a_temporary_folder, os.path.basename(s)) for s in local_index_files
+            ]
+            with pytest.raises(RuntimeError, match=f'.*Failed to download index.json.*'):
+                merge_index(index_file_urls, out, keep_local=keep_local)
+
+        with tempfile.TemporaryDirectory() as a_temporary_folder:
+            index_file_urls = [(os.path.join(a_temporary_folder, os.path.basename(s)), '')
+                               for s in local_index_files]
+            with pytest.raises(FileNotFoundError, match=f'.*Check data availability!.*'):
+                merge_index(index_file_urls, out, keep_local=keep_local)
+            return
+
+    if index_file_urls_pattern == 3:
+        remote_index_files = [
+            os.path.join(scheme, MY_BUCKET[scheme], MY_PREFIX, os.path.basename(o))
```

Comment:
Is this testing remote cloud path?

Reply:
yes, but no downloading, so still a local test.